zerver/views/message_send.py

```diff
@@ -339,3 +347,65 @@ def render_message_backend(
 
     rendering_result = render_markdown(message, content, realm=user_profile.realm)
     return json_success(request, data={"rendered": rendering_result.rendered_content})
+
+
+REPORT_MESSAGE_REASONS = ["spam", "harassment", "inappropriate", "norms", "other"]
+
+
+@has_request_variables
+def report_message_backend(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    message_id: int = REQ(converter=to_non_negative_int, path_only=True),
+    reason: str = REQ(default="other", str_validator=check_string_in(REPORT_MESSAGE_REASONS)),
+    explanation: Optional[str] = REQ(default=None),
+    # quoting is provided by client-side javascript, i.e. no server-side library
+    quoted_message: Optional[str] = REQ(default=None),
+) -> HttpResponse:
+    request_notes = RequestNotes.get_notes(request)
+    assert request_notes.log_data is not None
+    origmsg, ignored_user_message = access_message(user_profile, message_id, lock_message=False)
+    quoted_message = "" if quoted_message is None else quoted_message.strip()
+    explanation = "" if explanation is None else explanation.strip()
+
+    # todo: link/tooltip the reason to docs explaining what this means
+    rptmsg = f"{_('Reporting reason')}: **`"
+    if reason == "spam":
+        rptmsg += f"{_('Spam')}"
+    elif reason == "harassment":
+        rptmsg += f"{_('Harassment')}"
+    elif reason == "inappropriate":
+        rptmsg += f"{_('Inappropriate content')}"
+    elif reason == "norms":
+        rptmsg += f"{_('Violates community norms')}"
+    else:
+        rptmsg += f"{_('Other reason')}"
+        if explanation == "":
+            raise JsonableError(_("For reason=other, an explanation must be provided."))
+    rptmsg += f"`** by: @_**{user_profile.full_name}|{user_profile.id}** "
+    if explanation != "":
+        rptmsg += f" {_('who included these notes')}:\n```\n{explanation}\n```\n"
+
+    # put original msg at the end, so if rptmsg gets truncated, it's from this
+    rptmsg += "\n"
+    if origmsg.recipient.type == Recipient.STREAM:
+        stream = get_stream_by_id_in_realm(origmsg.recipient.type_id, user_profile.realm)
+        rptmsg += f"{_('Originally sent to ')}#**{stream.name}>{origmsg.topic_name()}**\n"
+    elif origmsg.recipient.type == Recipient.PERSONAL:
+        receiver = get_user_profile_by_id_in_realm(origmsg.recipient.type_id, user_profile.realm)
+        rptmsg += f"{_('Originally sent as a private message to')} @**{receiver.full_name}|{receiver.id}**\n"
+    else:
+        sentto = " ".join([f"@**|{id}**" for id in get_huddle_user_ids(origmsg.recipient)])
+        rptmsg += f"{_('Originally sent as a private message to')} {sentto}\n"
+    rptmsg += f"\n{_('Original message')}: {quoted_message}"
+
+    with override_language(user_profile.realm.default_language):
+        sender = get_system_bot(settings.NOTIFICATION_BOT, user_profile.realm_id)
+        topic = _("sent by") + " " + origmsg.sender.full_name
```

Comment:
We probably want the sender to be indicated by a mention; and also see the last comment about i18n. Probably we want something like:

```
Sent by {user}.
```

with `{user}` being silent mention syntax -- at least if we're not making this part of a more complex sentence.